optuna/trial/_trial.py

```diff
@@ -561,6 +564,12 @@ def objective(trial):
                 ``n_warmup_steps`` as the warmup mechanism.
         """
 
+        if len(self.study.directions) > 1:
+            _logger.warning(
+                "When the problem is multi-objective optimization, Trial.report does nothing."
+            )
+            return
```

Comment:
I'm wondering if we shouldn't just keep the previous behavior allowing pruning since the logic of `report`/pruning isn't currently affected by whether the study is SO or MO. It's clear from the parameters of `report` that it only takes a single value. Ditto for `should_prune`. I do think that some users will find it confusing trying to pass multiple "values" into the `value` argument, however. We might want to check this explicitly.

Reply:
Looks reasonable to me. We can use pruning with MO for one-dimensional values. It may be a specific coordinate of objectives in MO.

Reply:
> We can use pruning with MO for one-dimensional values.

I think we decided to keep the behavior of pruning undefined if the study is MO. This is mainly because we do not have almost any use cases, and it may be too early to provide the feature in any way. So, I personaly feel just raising the warning is reasonable for now.

Reply:
Sorry if I missed that consensus. Just thinking that pruning should conceptually be well defined even for MO studies since intermediate values and values are separately treated in Optuna. Couldn't one optimize model accuracy and latency, while pruning on say, either (or something entirely different)? As for visualization, you might still want to utilize `plot_intermediate_values` to visualize training progress of neural nets?

Reply:
> Just thinking that pruning should conceptually be well defined even for MO studies since intermediate values and values are separately treated in Optuna.

Hmm, I think we still have some design choices to store/utilize intermediate values for multi-objective studies.
As you mentioned, the current intermediate value handles only a single metric value, but we may want to support multiple metrics in the future. If we stop providing the feature, we may be able to keep the API clean. For this reason, I thought we decided not to provide the pruning feature for v2.4.0. I'm sorry if I'm wrong.

Reply:
[Just a note] Through an internal discussion, we have decided to do nothing for `trial.report` and `trial.should_prune` during multi-objective optimization. This is because we should balance the transferring cost from existing MO users and the migration cost in the future when MO pruning is available.

Reply:
[Just a note] From the continued internal discussion, we have decided to raise `NotImplementedError` for `trial.report` and `trial.should_prune` during multi-objective optimization. This is to make it easier for users to notice that pruning is not supported in MO.